kernel/src/syscall/signalfd.rs

```diff
@@ -0,0 +1,372 @@
+
+    if signal_file.mask().load(Ordering::Relaxed) != new_mask {
+        signal_file.update_signal_mask(new_mask)?;
+    }
+    signal_file.set_non_blocking(non_blocking);
+    Ok(fd)
+}
+
+fn register_observer(ctx: &Context, signal_file: &Arc<SignalFile>, mask: SigMask) -> Result<()> {
+    let filter = SigEventsFilter::new(mask);
+
+    ctx.posix_thread
+        .register_sigqueue_observer(signal_file.observer_ref(), filter);
+
+    Ok(())
+}
+
+bitflags! {
+    /// Signal file descriptor creation flags
+    struct SignalFileFlags: u32 {
+        const O_CLOEXEC = CreationFlags::O_CLOEXEC.bits();
+        const O_NONBLOCK = StatusFlags::O_NONBLOCK.bits();
+    }
+}
+
+/// Signal file descriptor implementation
+///
+/// Represents a file descriptor that can be used to receive signals
+/// as readable events.
+struct SignalFile {
+    /// Atomic signal mask for filtering signals
+    signals_mask: AtomicSigMask,
+    /// I/O event notifier
+    pollee: Pollee,
+    /// Non-blocking mode flag
+    non_blocking: AtomicBool,
+    /// Weak reference to self as an observer
+    weak_self: Weak<dyn Observer<SigEvents>>,
+}
+
+impl SignalFile {
+    /// Create a new signalfd instance
+    fn new(mask: AtomicSigMask, non_blocking: bool) -> Arc<Self> {
+        Arc::new_cyclic(|weak_ref| {
+            let weak_self = weak_ref.clone() as Weak<dyn Observer<SigEvents>>;
+            Self {
+                signals_mask: mask,
+                pollee: Pollee::new(),
+                non_blocking: AtomicBool::new(non_blocking),
+                weak_self,
+            }
+        })
+    }
+
+    fn mask(&self) -> &AtomicSigMask {
+        &self.signals_mask
+    }
+
+    fn observer_ref(&self) -> Weak<dyn Observer<SigEvents>> {
+        self.weak_self.clone()
+    }
+
+    fn update_signal_mask(&self, new_mask: SigMask) -> Result<()> {
+        if let Some(thread) = current_thread!().as_posix_thread() {
+            thread.unregister_sigqueue_observer(&self.weak_self);
+            let filter = SigEventsFilter::new(new_mask);
+            thread.register_sigqueue_observer(self.weak_self.clone(), filter);
+        }
+        self.signals_mask.store(new_mask, Ordering::Relaxed);
+        Ok(())
+    }
+
+    fn set_non_blocking(&self, non_blocking: bool) {
+        self.non_blocking.store(non_blocking, Ordering::Relaxed);
+    }
+
+    fn is_non_blocking(&self) -> bool {
+        self.non_blocking.load(Ordering::Relaxed)
+    }
+
+    /// Check current readable I/O events
+    fn check_io_events(&self) -> IoEvents {
+        if let Some(thread) = current_thread!().as_posix_thread() {
+            let mask = self.signals_mask.load(Ordering::Relaxed);
+            if thread.sig_pending().intersects(mask) {
+                IoEvents::IN
+            } else {
+                IoEvents::empty()
+            }
+        } else {
+            IoEvents::empty()
+        }
```

Comment:
Use `let.. else` to return early to avoid redundant indent.

```rust
let Some(thread) = current_thread!().as_posix_thread() else {
    return IoEvents::empty();
};

let mask = self.signals_mask.load(Ordering::Relaxed);
if thread.sig_pending().intersects(mask) {
    IoEvents::IN
} else {
    IoEvents::empty()
}
```